File: heap_sort.cpp
#include <iostream>
using namespace std;
void Heap(int arr1[], int x, int z)
{
	int largest_element = z;
	int j = 2 * z + 1;
	int k = 2 * z + 2;
	if (j < x && arr1[j] > arr1[largest_element])
		largest_element = j;
	if (k < x && arr1[k] > arr1[largest_element])
		largest_element = k;
	if (largest_element != z) {
		swap(arr1[z], arr1[largest_element]);
		Heap(arr1, x, largest_element);
	}
}
void heap_sort_function(int arr1[], int total_size_of_arr1)
{

	for (int x = total_size_of_arr1 / 2 - 1; x >= 0; x--)
		Heap(arr1, total_size_of_arr1, x);

	for (int x = total_size_of_arr1 - 1; x >= 0; x--)
        {
            swap(arr1[0], arr1[x]);
            Heap(arr1, x, 0);
        }
}
int main()
{
	int arr1[] = { 12, 11, 13, 5, 6, 7 };
	int total_size_of_arr1 = (sizeof(arr1)/sizeof(arr1[0]));
	heap_sort_function(arr1,total_size_of_arr1);
	for(int i=0;i<total_size_of_arr1;i++)
    {
        cout<<arr1[i]<<" ";
    }
	return 0;
}

File: heap_sort.py
def heap(Array, total_size_of_Array, z):
   
      largest_element = z
      left = 2 * z + 1
      right = 2 * z + 2
  
      if left < total_size_of_Array and Array[z] < Array[left]:
          largest_element= left
  
      if right < total_size_of_Array and Array[largest_element] < Array[right]:
          largest_element = right
  
      if largest_element != z:
          Array[z], Array[largest_element] = Array[largest_element], Array[z]
          heap(Array, total_size_of_Array, largest_element)
  
  
def heap_sort_function(Array):
      total_size_of_Array = len(Array)
  
      for x in range(total_size_of_Array//2, -1, -1):
          heap(Array, total_size_of_Array, x)
  
      for x in range(total_size_of_Array-1, 0, -1):

          Array[x], Array[0] = Array[0], Array[x]
  
          heap(Array, x, 0)
  
  
Array= [1, 12, 10, 15, 6, 19]
heap_sort_function(Array)
print("Sorted Elements: ")
print(Array)
